LIPS/lip-25.md

```diff
@@ -0,0 +1,1983 @@
+        _vettedKeysCountsOffset := add(calldataload(36), 36) // arg2 calldata offset + 4 (signature len) + 32 (length slot))
+    }
+    for (uint256 i; i < nodeOperatorsCount;) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            nodeOperatorId := shr(192, calldataload(add(_nodeOperatorIdsOffset, mul(i, 8))))
+            vettedKeysCount := shr(128, calldataload(add(_vettedKeysCountsOffset, mul(i, 16))))
+            i := add(i, 1)
+        }
+        _requireValidRange(nodeOperatorId < totalNodeOperatorsCount);
+        _updateVettedSingingKeysCount(nodeOperatorId, vettedKeysCount, false /* only decrease */);
+    }
+    _increaseValidatorsKeysNonce();
+}
+```
+
+Also, the existing `setNodeOperatorStakingLimit` external method and the  `_updateVettedSingingKeysCount` internal method should be updated.
+
+```solidity
+function setNodeOperatorStakingLimit(uint256 _nodeOperatorId, uint64 _vettedSigningKeysCount) external {
+    _onlyExistedNodeOperator(_nodeOperatorId);
+    _authP(SET_NODE_OPERATOR_LIMIT_ROLE, arr(uint256(_nodeOperatorId), uint256(_vettedSigningKeysCount)));
+    _onlyCorrectNodeOperatorState(getNodeOperatorIsActive(_nodeOperatorId));
+
+    _updateVettedSingingKeysCount(_nodeOperatorId, _vettedSigningKeysCount, true /* _allowIncrease */);
+    _increaseValidatorsKeysNonce();
+}
+
+function _updateVettedSingingKeysCount(
+    uint256 _nodeOperatorId,
+    uint256 _vettedSigningKeysCount,
+    bool _allowIncrease
+) internal {
+    Packed64x4.Packed memory signingKeysStats = _loadOperatorSigningKeysStats(_nodeOperatorId);
+    uint256 vettedSigningKeysCountBefore = signingKeysStats.get(TOTAL_VETTED_KEYS_COUNT_OFFSET);
+    uint256 depositedSigningKeysCount = signingKeysStats.get(TOTAL_DEPOSITED_KEYS_COUNT_OFFSET);
+    uint256 totalSigningKeysCount = signingKeysStats.get(TOTAL_KEYS_COUNT_OFFSET);
+
+    uint256 vettedSigningKeysCountAfter = Math256.min(
+        totalSigningKeysCount, Math256.max(_vettedSigningKeysCount, depositedSigningKeysCount)
+    );
+
+    if (vettedSigningKeysCountAfter == vettedSigningKeysCountBefore) return;
+
+    require(
+        _allowIncrease || vettedSigningKeysCountAfter < vettedSigningKeysCountBefore,
+        "VETTED_KEYS_COUNT_INCREASED"
+    );
+
+    signingKeysStats.set(TOTAL_VETTED_KEYS_COUNT_OFFSET, vettedSigningKeysCountAfter);
+    _saveOperatorSigningKeysStats(_nodeOperatorId, signingKeysStats);
+
+    emit VettedSigningKeysCountChanged(_nodeOperatorId, vettedSigningKeysCountAfter);
+
+    _updateSummaryMaxValidatorsCount(_nodeOperatorId);
+}
+```
+
+#### 3.3. New reward distribution engine
+
+The new external permissionless `distributeReward` method should be implemented. This method should distribute all accumulated module rewards among node operators based on the latest accounting report.
+
+Rewards can be distributed after node operators' statistics are updated until the next reward is transferred to the module during the next oracle frame.
```

Comment:
More input would be helpful here. The main motivation is to leave in the 3d phase of AO report only the update of operator data in modules to have more or less the same size of transactions

Reply:
Yes, agree with you. I've added much more detail about the new logic of the 3rd phase of the Accounting Oracle report to this section.